src/modm/driver/usb/stusb4500.hpp

```
@@ -0,0 +1,160 @@
+
+		current /= 10;
+		voltage /= 50;
+
+		buffer2.at(0) = current;
+		buffer2.at(1) = ((current >> 8) & 0b11) | ((voltage & 0b11'1111 ) << 2);
+		buffer2.at(2) = (buffer2.at(2) & 0b1111'0000) | ((voltage >> 6) & 0b1111);
+
+		RF_END_RETURN_CALL(updateRegister<4>(DpmSnkPdos.at(pdoNumber - 1), buffer2.data()));
+	}
+
+	/**
+	 * Sets the valid PDO.
+	 *
+	 * @param	pdoNumber Which PDO to set valid, range 1 to 3.
+	 */
+	modm::ResumableResult<bool>
+	setValidPdo(uint8_t pdoNumber)
+	{
+		RF_BEGIN();
+
+		if((pdoNumber < 1) || (pdoNumber > 3))
+			RF_RETURN(false);
+
+		RF_END_RETURN_CALL(updateRegister<1>(Register::DpmPdoNumb, &pdoNumber));
+	}
+
+	/**
+	 * Retrieve RDO status from STUSB4500 chip.
+	 *
+	 * @return	RdoRegStatusData object. Typically only
+	 *  RdoRegStatusData::MaxCurrent [mA] is of interest.
+	 */
+	inline modm::ResumableResult<RdoRegStatusData>
+	getRdoRegStatus()
+	{
+		RF_BEGIN();
+
+		RF_CALL(readRegister<4>(Register::RdoRegStatus, buffer2.data()));
+
+		// MaxCurrent: Bits 9..0; OperatingCurrent: Bits 19..10; ObjectPos: Bits 30..28
+		rdoRegStatusData.MaxCurrent = buffer2.at(0) | ((buffer2.at(1) & 0b11) << 8);
+		rdoRegStatusData.MaxCurrent *= 10;
+		rdoRegStatusData.OperatingCurrent = ((buffer2.at(1) & 0b1111'1100) >> 2) | ((buffer2.at(2) & 0b1111) << 6);
+		rdoRegStatusData.OperatingCurrent *= 10;
+		rdoRegStatusData.ObjectPos = (buffer2.at(3) & 0b0110'0000) >> 5;
+
+		RF_END_RETURN(rdoRegStatusData);
+	}
+
+private:
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	readRegister(Register reg, uint8_t* output)
+	{
+		RF_BEGIN();
+		buffer[0] = uint8_t(reg);
+		this->transaction.configureWriteRead(buffer.data(), 1, output, length);
+		RF_END_RETURN_CALL( this->runTransaction() );
+	}
+
+	template<uint32_t length>
+	modm::ResumableResult<bool>
+	updateRegister(Register reg, uint8_t* data)
+	{
+		static_assert(length <= (std::tuple_size<decltype(buffer)>::value - 1), "buffer too small");
```

Pass length as an argument and use `length = std::min(length, 5);`

---

A span would be appropriate. And why use tuple_size. std::array is constexpr.

`static_assert(4 <= (buffer.size() - 1), "buffer too small");`

---

Nope. `array.size()` does not work in static_assert() because `buffer` is not constexpr. Try it out yourself...

---

> use `length = std::min(length, 5);`

No, because I don't like silent failures. Run-time checking the `length` parameter and abort with `return false` now.